### DISCUSSION TITLE
How to Indent Wrapped Text?

### DISCUSSION THREAD

**Discussion by KyleKing** [Textualize/rich#2792] [Q&A]
How do I add indents to wrapped lines? I tried overriding `Text.wrap` and looking through a few of the Console methods, but there didn't seem to be a configuration option.

I would like to add indents to the wrapped text when printed, so that:

```py
from rich.console import Console
from rich.text import Text

text = Text()
text.append('key: ---------- ' * 50)
Console().print(text)
```

Outputs something like this (where each new line has an extra four spaces):

```sh
key: ---------- key: ---------- key: ---------- key: ---------- key: ---------- key: ---------- key: ---------- key: ---------- key: ---------- key: ---------- key: ---------- key: ---------- key: ---------- key:
    ---------- key: ---------- key: ---------- key: ---------- key: ---------- key: ---------- key: ---------- key: ---------- key: ---------- key: ---------- key: ---------- key: ---------- key: ---------- key:
    ---------- key: ---------- key: ---------- key: ---------- key: ---------- key: ---------- key: ---------- key: ---------- key: ---------- key: ---------- key: ---------- key: ---------- key: ---------- key:
    ---------- key: ---------- key: ---------- key: ---------- key: ---------- key: ---------- key: ---------- key: ---------- key: ---------- key: ---------- key: ----------
```

I tried `end`, but that is only after the line and not on fold:

```py
console.print(text, end="\n    ")
```

Blocking: https://github.com/KyleKing/tail-jsonl/pull/1

**Comment by harkabeeparolus** [comment]
This is usually known as a **hanging indent** in word processing. The trick is that the first line is not indented at all, whereas all the subsequent lines are.

I really want this feature too! 😳

You can of course indent all wrapped lines by the **same** amount using e.g. [`Padding.indent()`](https://rich.readthedocs.io/en/stable/reference/padding.html#rich.padding.Padding.indent), but in this case we want a different indent for the first line.

I wish `Padding.indent()` had an optional parameter to specify the indent for the first line, perhaps "first_line_level". In that case, you could implement a hanging indent by doing something like:

```python
hanging = Padding.indent(Text("key: ---------- " * 50), level=4, first_line_level=0)  # TODO: Does not exist :-(
```

**Reply by harkabeeparolus** [reply]
I kept searching through the discussions here... It should be possible to implement hanging indents using [the custom class](https://github.com/Textualize/rich/discussions/1686#discussioncomment-1639502) in discussion #1686. If I have the time, I'll look into it.

**Reply by harkabeeparolus** [reply]
This kind of works:

```python
"""Indent Rich text or renderable."""

from dataclasses import dataclass
from typing import Optional

from rich.abc import RichRenderable
from rich.console import Console, ConsoleOptions, RenderResult
from rich.padding import Padding
from rich.segment import Segment


@dataclass
class Indent:
    """Indent a Rich renderable, with a different indent for the first line."""

    renderable: RichRenderable
    level: int
    first_line_level: Optional[int] = None

    def __rich_console__(
        self, console: Console, options: ConsoleOptions
    ) -> RenderResult:
        extra = max(self.level, self.first_line_level or 2)
        padded = Padding(self.renderable, (0, extra, 0, 0))
        segments = console.render(padded, options)
        lines = Segment.split_lines(segments)
        printing_first_line = True
        indent = (
            self.first_line_level if self.first_line_level is not None else self.level
        )
        for line in lines:
            yield Segment(" " * indent)
            yield from line
            yield Segment("\n")
            if printing_first_line:
                printing_first_line = False
                indent = self.level


if __name__ == "__main__":
    from rich.table import Table
    from rich.text import Text

    con = Console()
    lorem = Text(
        "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod "
        "tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, "
        "quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo "
        "consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse "
        "cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non "
        "proident, sunt in culpa qui officia deserunt mollit anim id est laborum."
    )
    hanging = Indent(lorem, level=10, first_line_level=0)
    con.rule("Normal printing")
    con.print(hanging)
    con.line(2)
    table = Table("Normal", "Hanging")
    table.add_row(lorem, hanging)
    con.print(table)
```

The first line is too short, but

**Reply by KyleKing** [reply]
Wow, that is exactly what I was looking for!

**Reply by harkabeeparolus** [reply]
Oh shoot. I just noticed that it doesn't actually work; the indented version silently crops characters at the right margin. 😣 It does make sense, though, because the line lengths are rendered before the indent is added. I edited my code above to extend the right margin with `Padding()`, but then the first line is too short so it doesn't look very nice.

Well, that sucks. I'll have to go back to the drawing board. 😕

**Reply by willmcgugan** [reply]
Can't think of a quick way of implementing that using the existing API. Sounds a bit like an XY problem though. Why do you need it?

**Reply by KyleKing** [reply]
I wanted to add hanging indents to my log formatting tool: https://github.com/KyleKing/tail-jsonl

Since I initially posted this discussion, I've tweaked the colors and styling to make the JSONL lines easier to visually scan, which generally seems to work well.

However, for non-JSONL text, I pass those lines through without styling, which could benefit from hanging indents, though it's not something I absolutely need, and I'm happy enough with how the output looks today

**Reply by harkabeeparolus** [reply]
Yeah, printing log lines, email headers or similar text info but with word wrapping is my primary use case too. 👍🏻

Hanging indents are normally used for a bibliography or reference section. But maybe that's more a topic for word processing and not something you would do in the terminal. 😊
